### DISCUSSION TITLE
Enhance shadows and brightness

### DISCUSSION THREAD

**Discussion by VivitionDeveloper** [libvips/libvips#4036] [Help wanted]
I've played around with vips_tonelut and vips_maplut, but it does not have the results wanted. What we like to achieve is:

- see better differences between blacks/shadows. In the example photo: the differences between floor and tyres are poor now.
- address the too low level of light/brightness. The lights in the studio do either suffer from age (=become less bright over time) or too less light (just due to lack of enough LED tubes, or too much absorption by the floor or wall). Cameras are set to manual exposure and those settings do not change over time.
- can HDR play a role here and how to achieve HDR with libvips?
- does it make sense to apply tone mapping to LCH to the 'L' channel only? And how to use this in combination with vips_tonemap()? Do I need to use vips_tonemap() on the sRGB JPEG or first perform vips_bandsplit, vips_tonemap, vips_maplut on channel 0 and then vips_bandjoin, vips_colorspace(sRGB) and then save the result? Or does the used color space not play a role here?
- any other advices are welcome, e.g the order of vips operations, e.g. regarding previous operations + sharpening (vips_sharpen())

I've added an example image and played around with NIP2, but did not get satisfying results (even adjust tone curve comes with somewhat disappointing result). When needed, I can add that result as well, but as I'm not experienced with image (light, shadow,) enhancements, I may have performed the wrong actions (for now: just plain adjust tone mapping on a sRGB image). 
![sample](https://github.com/libvips/libvips/assets/8791449/b1252d3c-f61b-4b24-9e15-a7f1276e7f17)

I also discovered that adjusting the tone curve, and setting shadow adjust to e.g. 7, the red color get less saturized (i.e. the color of the car changes from the real color). In ideal case, the colors should remain (almost) the same. Is it possible to adjust shadows/brightness without losing color saturation?

Please let me know your recommendations.

BTW we are already using libvips for thumbnails, resizes, crops, sharpening etc.

**Comment by VivitionDeveloper** [comment]
Small correction: vips_tonemap() was misspelled, but is vips_tonelut().

**Comment by jcupitt** [comment]
Hello @CoachRDeveloper,

Usually you'd transform to CIELAB, tonemap the L, and transform back to RGB. nip2 will to do this automatically I think. Perhaps:

![image](https://github.com/libvips/libvips/assets/580843/e2066279-3fd6-4427-bf65-15e6df87f0fb)

But of course it depends on your needs. The libvips tonemap operation was originally designed for prepress work.

Generally you'd hire a good photographer and they'll do all this for you in lightroom or the phase one thing.

**Comment by VivitionDeveloper** [comment]
Many thanks for reply and providing the information. As taking photos is automated and cameras preinstalled, hiring a good photographer is not realistic. Usually it is lack of willingness to invest in enough LED tubes or changes made by the owner of the studio to e.g. the floor (e.g. switching to a black floor) without adjusting the required amount of lights. Therefore, we have to rely on some tricks to enhance the photo quality to some extend.

From what I tried myself (looks much like you did with tonelut), I noticed that the red color became less vibrant when enhancing the shadows.

But which operation did you apply to enhance the photo? I have to do it programmatically in C++.
Is adjusting the brightness not a good way to reduce the flooded black shadows?
Or is HDR something that is useful to address this?

Please forgive me my lack of knowledge about when to apply enhancing brightness, tone curve or HDR (with the use of libvips, as our code already is using libvips).

**Comment by VivitionDeveloper** [comment]
I forgot something: is there something as 'a generic tone mapping' curve (=LUT table) that can be applied to any car (under same lightness conditions, taken from same location), even if the car color is different (which can be any color)?

And how to address aging of the LED tubes (i.e. reduced light output).?

**Comment by jcupitt** [comment]
The tone map menu item in nip2 doesn't change the saturation or hue, just the brightness. If the colours seem less vibrant, it may be your display. 

Accurate colour matching is a painful and expensive business -- you need a high-quality display, you need to generate a custom display profile, you need to standardise viewing conditions in the room (ie. paint the walls with a good-quality neutral 50% grey paint, use high-quality lighting, exclude daylight, wait for 10 minutes after entering the room and before starting work, have a viewing cabinet for the match item, etc.). I spent years doing this :( 

Yes, `vips_tonelut()` makes a tone curve from the parameters that you can apply to any photo. I would:

- transform to 16-bit CIELAB
- extract the L band (the first)
- take the histogram and find the 1% and 99% points (ie. exclude specular highlights and 0s in the image)
- generate a tone curve using the 1% and 99% points for `Lb` and `Lw`
- map the L band, then reattach the AB bands from the original
- transform back to sRGB

https://www.libvips.org/API/current/libvips-create.html#vips-tonelut

**Reply by VivitionDeveloper** [reply]
Thanks. I have some C++ code ready now, but the result differs. Am I doing something wrong?

Code (with some wrapper code of ourselves regarding string/file manipulation, but that should be self-explaining):
```
	GFile flInput(GetInput("Input file"));
	vips::VImage vipsImage = vips::VImage::new_from_file(flInput.c_str());
	vipsImage = vipsImage.colourspace(VipsInterpretation::VIPS_INTERPRETATION_LAB);
	auto vipsLUT = vipsImage.tonelut(vips::VImage::option()->
		// need to set max values, otherwise tonelut seems not to do much...
		set("in-max", 255)->set("out-max", 255)->
		set("Lb", 0)->set("Lw", 100)->
		set("Ps", 0.2)->set("Pm", 0.5)->set("Ph", 0.8)->
		set("S", 15)->set("M", 30)->set("H", 15));
	auto vipsChannels = vipsImage.bandsplit();
	// tonemap the 'Luminance' part
	vipsChannels[0] = vipsChannels[0].maplut(vipsLUT);
	vipsImage = vipsChannels[0].bandjoin(vipsChannels[1]).bandjoin(vipsChannels[2]);
	vipsImage = vipsImage.colourspace(VipsInterpretation::VIPS_INTERPRETATION_sRGB);
	vipsImage.write_to_file(GString("C:\\Temp\\Kees\\" + flInput.getBaseName() + "_out.jpg[Q=95,optimize_coding,strip]").c_str());
```

Resulting image attached. 
![sample_out](https://github.com/user-attachments/assets/f3544a4c-c423-4c3c-a3c7-57f88b14e2cf)

Your result was much better, however, I was not able to map nip2 actions to libvips operations (and made may C++ counterpart).

**Reply by kmartinez** [reply]
oh maybe LABQ and even LABS could be deprecated some day - and cleared out - as we made it to save on disk space years ago. However I bet there are LabS/Q images in an archive somewhere!
our demo images certainly have some: http://marc.ecs.soton.ac.uk/vips/

**Comment by VivitionDeveloper** [comment]
It is somewhat unclear to me what the difference is between VIPS_INTERPRETATION_LABS and VIPS_INTERPRETATION_LAB. I guess the latter is 8-bit, although not explicitly mentioned on https://www.[libvips](https://www.libvips.org/API/current/VipsImage.html).org/API/current/VipsImage.html, but for VIPS_INTERPRETATION_LABS, I don't need to set in-max/in-min, as the default (short int values) seem to suffice.

**Comment by jcupitt** [comment]
LABS is 16-bit (ie. short) CIELAB, LABQ is a strange 8 bit CIELAB with some extra bits in the fourth byte. LAB is anything you like, but pixels need to be 0-100 for L and +/-128 for AB.

https://www.libvips.org/API/current/VipsImage.html#VipsInterpretation

I played about with your image a little more and I think the main problem is the underexposure. I would go to XYZ first and scale everything up by maybe a factor of 2.5:

![x](https://github.com/user-attachments/assets/033d797f-3eb4-475d-b6a6-41397190dd54)

If you need to lift the tyres a bit more than that, you could just boost the shadows:

![x2](https://github.com/user-attachments/assets/82ce9ec7-9bfe-420e-be2e-c1f345af1528)

**Reply by jcupitt** [reply]
You could maybe alter the colour temperature a bit too, it's rather creamy right now.

**Comment by jcupitt** [comment]
I made a tiny demo prog for you:

```C++
/* compile with
 *  g++ -g -Wall tonemap.cpp `pkg-config vips-cpp --cflags --libs`
 */

#include <vips/vips8>

using vips::VImage;

int
main(int argc, char **argv)
{
    if (VIPS_INIT("")) 
        vips_error_exit(nullptr);

    auto image = VImage::new_from_file(argv[1]);

    // import using the embedded profile, and import to XYZ PCS
    image = image.icc_import(VImage::option()
        ->set("embedded", TRUE)
        ->set("pcs", "xyz"));

    // brightness boost ... this is best done in XYZ space, where pixel values
    // are proportional to the number of photons
    //
    // 2.5 seems about right for the sample image (bad underexposure)
    image *= 2.5;

    // to labq, then labs .. this route clips out of range values
    // (with 8.16 you can go straight to labs)
    auto labs = image
        .colourspace(VIPS_INTERPRETATION_LABQ)
        .colourspace(VIPS_INTERPRETATION_LABS);
    auto l = labs[0];
    auto ab = labs.extract_band(1, VImage::option()->set("n", 2));

    // L will be ushort with L* scaled to fill 0-32767 
    //
    // take the histogram and find the 1% and 99% points, then set that as the
    // range ... we want to avoid specular highlights and any pools of 0s
    auto black = l.percent(1.0);
    auto white = l.percent(99.0);

    printf("black = %d\n", black);
    printf("white = %d\n", white);

    // make a tone curve for this image
    auto tone = VImage::tonelut(VImage::option()
        // tonal range
        ->set("Lb", 100.0 * black / 32767.0)
        ->set("Lw", 100.0 * white / 32767.0)

        // position of shadow, midtone and highlight in 0-1
        ->set("Ps", 0.15)
        ->set("Pm", 0.5)
        ->set("Ph", 0.8)

        // shadow, midtone and highlight adjust, +/- 30
        //
        // shadows +10 seems to help the tyres
        ->set("S", 10.0)
        ->set("M", 0.0)
        ->set("H", 0.0)
    );

    // use the tonecurve to adjust L*
    l = l.maplut(tone);

    // rebuild the labs image, then back to srgb
    image = l
        .bandjoin(ab)
        .copy(VImage::option()->set("interpretation", VIPS_INTERPRETATION_LABS))
        .colourspace(VIPS_INTERPRETATION_sRGB);

    image.write_to_file(argv[2]);

    return 0;
}
```

It makes:

![x](https://github.com/user-attachments/assets/0d9a8556-5162-4e3b-b8a3-5fc608f84ab3)

**Comment by VivitionDeveloper** [comment]
Wow, many thanks. And tricks like using LABQ before moving to LABS, I could never have invented this myself. I will try your code today. Again, thanks for great help.

The color temperature is typically determined by the camera, when adjusting the white balance upon first shoot in the car studio.
 
I hope I don't bother you with some other questions:

Q1:
Will increasing the brightness only (the L channel of LABS) with e.g. a fixed value result in a similar result, or will playing with tone mapping provide better results? I do not know which technique saves most details in a (JPEG) image (our source is always JPEG, straight out of the camera). I'm a C++ programmer, not a photographer or designer, sorry....

Q2:
With 
`vips::VImage vipsImage = vips::VImage::new_from_file(flInput.c_str());`
Then
`const auto stats = vipsImage.stats();`
Returns the statistics, but then rows are the y part I guess? in:
`auto agvAll = stats(4, 0);`

The documentation https://www.[libvips](https://www.libvips.org/API/current/libvips-arithmetic.html#vips-stats).org/API/current/libvips-arithmetic.html#vips-stats says that row 0 is for all bands, and then 1 and so on per band. However, in the getter https://www.libvips.org/API/8.11/cpp/classVImage.html, search for operator(), then it is not described wether x points to rows or columns. It seems to be rows. Can you please confirm? 

I have checked this, because the following lines return the same value, which is hopefully a proof of the previous conclusion:
```
auto agvAll = stats(4, 0);
auto avg = vipsImage.avg();
```

Q3: Is there a possibility to let show/trace NIP2 the executed VIPS commands? I guess that vips.exe is being called and with tracing, I would have more information to deduce the required C++ actions (like you now did for me, but I don't like to bother you each time)....

Q4: I just discovered that 8.16 is not yet released. I'm at 8.15.1 (just one version behind last released one), on Windows. When will 8.16 get released?

**Comment by VivitionDeveloper** [comment]
I know the problem is underexposure. But the owners of the car studio have their own reasons for *not* upgrading their studio with additional LED tubes. That's why we have to apply some 'tricks' to provide better quality, as underexposed photos are not an advertisement for us. So we have to do something, but still need to convince the studio owner that uprading the exposure conditions are the better approach (which is not easy).

**Comment by VivitionDeveloper** [comment]
I've tried out your suggestions and the result is very good. The biggest enhancement is in adjusting the brightness and tonemapping enhances the details in the tyres. Thanks again.

**Comment by jcupitt** [comment]
Underexposure is best fixed by multiplying by a constant in XYZ space. 

Underexposure is a lack of photons (obviously), XYZ space has numeric values proportional to the number of photons, so just scaling up is a very easy fix. 

You generally want to get the brightest diffuse reflector in the scene (the left-hand back wall in this case) up to about Y 98. The specular highlights on the car (eg. paint, chrome) can go quite a bit higher than this (they reflect the lights directly back into the camera), so you need to be a little careful what you set exposure off.

If they are happy with the colour balance, then of course leave it, but you would generally aim for the media white point, which is 250, 250, 250 for JPEG, and they are some way off that.

**Comment by jcupitt** [comment]
Yes, (x, y) is always (columns, rows), like in maths.

You can right-click on an image in nip2 and select History to get a list of the commands that generated that image. It's using the vips7 API though, so translating to vips8 isn't always easy. I'm working on a vips8 update for nip2 (will be nip4 heh) at the moment:

https://github.com/jcupitt/nip4

8.16 will be some time late this summer, perhaps August or September.

**Reply by VivitionDeveloper** [reply]
Thanks again for your reply. I am going to proceed with creating a simple way to apply similar logic to a (large) series of photos tailored to studio lighting conditions. Your answers were a great help.

**Reply by VivitionDeveloper** [reply]
> 8.16 will be some time late this summer, perhaps August or September.

Can you please confirm whether 8.16 indeed allows to move (from RGB) to VIPS_INTERPRETATION_LABS color space in one step, i.e. without the need of an intermediate step of switching to VIPS_INTERPRETATION_LABQ first?

**Reply by jcupitt** [reply]
Yes, I think so. Make a bug report if it fails.

**Comment by VivitionDeveloper** [comment]
One small addition: Because we are typically reading the images from a SMB share, I have played around with reading in sequential mode:
`auto image = VImage::new_from_file(argv[1], VImage::option()->set("access", VIPS_ACCESS_SEQUENTIAL));`

However, then the above example crashes at the end, when dereferencing memory (on Windows, using libvips 8.15.1).
The crash is in ViImage.cpp, call_option_string(), upon check of vips_cache_operation_buildp. It throws an error.
This error does not occur when reading in the default (random access) mode.

In our code, the error is slightly different: adjusting the brightness worked fine, i.e. after the line:
` image *= 2.5;`

I was able to save the file, but when executing:
`auto black = l.percent(1.0);`

then I got the following warnings + error:
```
(process:20052): VIPS-WARNING **: 11:09:23.036: error in tile 0 x 0
(process:20052): VIPS-WARNING **: 11:09:23.038: error in tile 0 x 8
VipsJpeg: out of order read at line 4000
```
The latter was caused by an exception. I don't understand why I got in our code another error, as that looks identical to your code (but using our own wrapper class and put adjust brightness and change lumininance in two distinct methods). I was not able to reproduce this now in a smaller example. Also, the crash in the small example was also not reproducible in our own code (unfortunatelly).

When I use the default open mode, then no issues happened.  Also VIPS_ACCESS_SEQUENTIAL_UNBUFFERED causes the same error.
Not sure whether I've to open another issue for this, but apparently sequential reads causes strange things.

But the final question (and the reason why I tried this): does SMB reads (whith a NAS behind, with traditinal HDDs) benefit from using libvips sequential reads (of course, at the price of using more memory)?

**Reply by jcupitt** [reply]
Yes, this example needs random access mode -- it uses `percent()` to find the 1 and 99 percent points. If you use hardcoded values instead, it'll run in seq mode.

It's not really a crash, just an uncaught exception.

**Reply by jcupitt** [reply]
Did you see:

https://www.libvips.org/API/current/How-it-opens-files.html

It explains how sequential mode works and why you had the error.

**Reply by VivitionDeveloper** [reply]
Many thanks for reply. That is useful information. I don't really understand why the percent() function mismatches with sequential read, but I've now noticed it...

**Reply by jcupitt** [reply]
With sequential images, you can only make one top-bottom scan of the input when writing the output.

The first `percent()` operation does a scan of the image to build a histogram, then computes the percent from that. The second `percent()` will try to scan the image again, but the first percent has already used the whole image up, so there's nothing left :(

If you remove the percent operations, it'll work fine, of course, but you won't have the autoranging.

**Reply by VivitionDeveloper** [reply]
Thanks a lot for clarification. It is a bit tricky that the operations on a image affect how the image must be opened. I did not expect such. I.e. when an image is sequentially read into memory (at the price of using more memory) then I would expect not to see different behavior. That's probably something to mention somewhat more explicitly in the documentation. However, that's a very small remark from my side. I'm happy with the detailed and very useful answers of you.

If it is not too much work for you, another question popped up. If you have time, I really like to get an explanation on the copy line in the code below. Is that line required? Or is it unnessesary? Or what is the purpose of adding that line?

 ```
image = l
        .bandjoin(ab)
        .copy(VImage::option()->set("interpretation", VIPS_INTERPRETATION_LABS))
        .colourspace(VIPS_INTERPRETATION_sRGB);
```

**Reply by jcupitt** [reply]
Sequential mode is a hint to the image open system that this pipeline will only make a single pass over the pixels (ie. "I will only need sequential access"). If you don't set it, then libvips assumes you want random access.

Imagine you are reading an image from a pipe, perhaps it's coming over a connection from a bucket on S3. You can only read the data once since you can't `seek()` pipes.

If you set sequential mode, libvips will connect your image pipeline directly to the pipe, so you'll process the image pixel by pixel as it streams in from S3. The image will never be buffered in memory.

If you don't, then libvips will load the entire image into RAM, then run your pipeline on the memory buffer. You can do more complex operations, but you'll need to allocate a huge memory buffer, and processing can't start until the whole thing is read in.

**Reply by jcupitt** [reply]
The `copy()` operation has no runtime cost, it just copies a pointer to make the new image, so don't worry about efficiency.

The new image it makes has the `interpretation` tag changed, that is, the pixels don't change, just the way libvips interprets their values. In this case, we're tagging them as 16-bit CIELAB. The `colourspace()` operation uses this to pick the best set of conversions to make an sRGB output for you.

**Reply by jcupitt** [reply]
... imagine a 16-bit image, and pixel (0, 0) has the value `[23562, 123, 2666]`. How should that be converted to sRGB?

To do this, you need to know how this image encodes colour values. For LABS, it's 0 .. 32767 for the first value (L*), and -32768 .. 32767 for the second two (a\*b\*), so that colour would be a blueish medium grey, perhaps `[121, 125, 140]` in sRGB (just a guess).

**Reply by VivitionDeveloper** [reply]
So, if I understand right: keep a copy as it does not harm? 
As a general rule when moving to another colourspace: it is required to add a copy too?
It libvips requires to do so, I'm a bit surprised that it is not mentioned e.g. at [vips_colourspace](https://www.libvips.org/API/current/libvips-colour.html#vips-colourspace.org/API/current/libvips-colour.html#vips-colourspace)
But my lack of knowledge is probably the main cause...

I really get confused when another part of the provided code does not perform such copy, but do not understand why:

```
 auto labs = image
        .colourspace(VIPS_INTERPRETATION_LABQ)
        .colourspace(VIPS_INTERPRETATION_LABS);
```

**Reply by jcupitt** [reply]
After the `bandjoin()`, the interpretation will not be set correctly, it'll be `grey16`, which is completely wrong. The program needs to tell libvips how to interpret the pixel values.

Try adding:

```C++
    // use the tonecurve to adjust L*
    l = l.maplut(tone);

    image = l
        .bandjoin(ab);
    image.write_to_file("x.v");

    // rebuild the labs image, then back to srgb
    image = l
        .bandjoin(ab)
        .copy(VImage::option()->set("interpretation", VIPS_INTERPRETATION_LABS))
        .colourspace(VIPS_INTERPRETATION_sRGB);
```


And looking at `x.v` in `vipsdisp`. You'll see a crazy-looking image.

**Reply by VivitionDeveloper** [reply]
Thanks for this information.

**Comment by VivitionDeveloper** [comment]
In addition to my previous question today (see above) about 8.16, some minor clarification and a reference to the piece of code supplied by you that triggers this question: Does 8.16 indeed solve the issue mentioned in this code piece (see also above for the whole part/contect)

```
    // to labq, then labs .. this route clips out of range values
    // (with 8.16 you can go straight to labs)
    auto labs = image
        .colourspace(VIPS_INTERPRETATION_LABQ)
        .colourspace(VIPS_INTERPRETATION_LABS);
    auto l = labs[0];
    auto ab = labs.extract_band(1, VImage::option()->set("n", 2));

**Reply by kleisauke** [reply]
IIUC, this was fixed with commit 088330c112ae15cfe44af1202149b6cf050174d1. So yes, 8.16.0 has this fix.

**Reply by VivitionDeveloper** [reply]
Thanks, as it quite difficult for me to check the source code and even see (the absence of) clipping. I will try out the new version (most probably this week). Of course, I do not expect bugs ;-).
